### PR TITLE
Refactor ciao certificate generation

### PR DESCRIPTION
**Pull request by rbradford**
This PR forms part of the pre-work of a simpler deployment methodology. Through the commits in this PR the code for creating certificates is refactored out from ciao-cert and into the ssntp/certs package.

### CODE REVIEW

**Comment by rbradford** [comment]
@sameo Is this API close to what you had in mind?

**Comment by sameo** [comment]
@rbradford Yes, this looks really good.

File: ssntp/certs/certs.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	var priv interface{}
+	var err error
+
+	if ell == false {
+		priv, err = rsa.GenerateKey(rand.Reader, 2048)
+	} else {
+		priv, err = ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	}
+
+	if err != nil {
```
Comment (markdryan) [inline ssntp/certs/certs.go]: Maybe we could make this function simpler.  If GenerateKey always sets priv to nil when there is an error, which it really should, we could just write.

```
func generatePrivateKey(ell bool) (interface{}, error) {
      if ell == false {
		return rsa.GenerateKey(rand.Reader, 2048)
	} else {
		return ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
	}
}
```

File: ssntp/certs/certs.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	// Create certificate signed by private key from serverCert
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, parentCert, publicKey(priv), serverPrivKey)
+	if err != nil {
+		return fmt.Errorf("Unable to create client certificate: %v", err)
+	}
+
+	// Write out certificate (including private key)
+	pem.Encode(certOutput, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
```
Comment (markdryan) [inline ssntp/certs/certs.go]: We should probably check the error return by Encode here.

File: ssntp/certs/certs.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, template, publicKey(priv), priv)
+	if err != nil {
+		return fmt.Errorf("Unable to create server certificate: %v", err)
+	}
+
+	// Write out CA cert
+	pem.Encode(caCertOutput, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+
+	// Write out certificate (including private key)
+	pem.Encode(certOutput, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
```
Comment (markdryan) [inline ssntp/certs/certs.go]: We need to check the error here.

File: ssntp/certs/certs.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+//    golang/src/pkg/crypto/tls/generate_cert.go
+//
+// which is:
+//
+//    Copyright 2009 The Go Authors. All rights reserved.
+//    Use of this source code is governed by a BSD-style
+//    license that can be found in the golang LICENSE file.
+//
+
+package certs
```
Comment (markdryan) [inline ssntp/certs/certs.go]: I think we're going to need some unit tests for all this stuff, particularly as we are adding a new package.

File: ssntp/certs/certs.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+func pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
+	case *ecdsa.PrivateKey:
+		b, err := x509.MarshalECPrivateKey(k)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to marshal ECDSA private key: %v", err)
+			os.Exit(2)
```
Comment (rbradford) [inline ssntp/certs/certs.go]: I fixed a bunch of places in the code to return errors. Look like I missed this one. Fixed!

File: ssntp/certs/certs.go (hunk 1 above)
Comment (rbradford) [inline ssntp/certs/certs.go]: Fixed!

File: ciao-cert/main.go [hunk 6]
```diff
@@ -202,138 +122,58 @@ func checkCompulsoryOptions() {
...
-			serverPrivKey, err = x509.ParsePKCS1PrivateKey(privKeyBlock.Bytes)
-		} else {
-			serverPrivKey, err = x509.ParseECPrivateKey(privKeyBlock.Bytes)
-		}
+		err = certs.CreateClientCert(template, *isElliptic, bytesCert, certOut)
 		if err != nil {
-			log.Fatalf("Could not get server private key %s", err)
+			log.Fatalf("Failed to create certificate: %v", err)
 		}
+		certOut.Close()
```
Comment (rbradford) [inline ciao-cert/main.go]: Ditto!

File: ssntp/certs/certs.go (hunk 2 above)
Comment (rbradford) [inline ssntp/certs/certs.go]: All pem.Encode returns now error checked.

**Comment by markdryan** [comment]
Passed Single VM verify.sh and BAT tests.

**Pull request merged**